parser-typechecker/src/Unison/Codebase/Editor/Propagate.hs

```diff
@@ -72,6 +75,27 @@ propagateAndApply patch branch = do
   f     <- applyPropagate edits
   pure $ Branch.step (f . applyDeprecations patch) branch
 
+-- Creates a mapping from old data constructors to new data constructors
+-- by looking at the original names for the data constructors which are
+-- embedded in the Decl object because we carefully planned that.
+generateConstructorMapping
+  :: Eq v
+  => Map v (Reference, Decl v _)
+  -> Map v (Reference, Decl.DataDeclaration' v _)
+  -> Map
+       (Reference, Int, ConstructorType)
+       (Reference, Int, ConstructorType)
+generateConstructorMapping oldComponent newComponent = Map.fromList
+  [ let t = Decl.constructorType oldDecl in ((oldR, oldC, t), (newR, newC, t))
+  | (v1, (oldR, oldDecl)) <- Map.toList oldComponent
+  , (v2, (newR, newDecl)) <- Map.toList newComponent
+  , v1 == v2
```

Comment:
I think it's fine given the relatively small number of constructors in data types written by humans, but this is O(n^2) when it could be O(n log n) - just iterate through the first map and do lg n lookups into the second map for each entry. There's an [`intersectionWith` function on `Map`](https://hackage.haskell.org/package/containers-0.6.2.1/docs/Data-Map-Internal.html#v:intersectionWith).